### PR TITLE
ClusterAutoscaler: Put APIs in a separate go module

### LINKED ISSUE #6307: CA: Put APIs in a separate go module
**Which component are you using?**:

ClusterAutoscaler

**Is your feature request designed to solve a problem? If so describe the problem this feature should solve.**:

If the external project imports cluster-autoscaler, the external project can't run `go list -modfile=./go.mod -m all`
because the cluster-autoscaler depends on k/k repository.

By these limitations, Goland (JetBrains) wouldn't work fine.
Please see the details: https://github.com/kubernetes-sigs/kueue/issues/1345

**Describe the solution you'd like.**:

As we discussed in the above issue with @x13n and @alculquicondor, the best solution is to put APIs in a separate go module.

**Describe any alternative solutions you've considered.**:

**Additional context.**:

### PR DESCRIPTION
**Pull request by tenzen-y**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
I put APIs in a separate go module.

#### Which issue(s) this PR fixes:

Fixes #6307

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by tenzen-y** [comment]
Also, it might be worth it to create a separate go module in https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/provisioningrequest/client as well.

@x13n WDYT?

**Comment by alculquicondor** [comment]
I don't think it's worth maintaining 3 packages. Maybe the client and APIs can be in the same package?

File: cluster-autoscaler/provisioningrequest/apis/go.mod
Comment (alculquicondor) [inline cluster-autoscaler/provisioningrequest/apis/go.mod:1]: I think the module should be
`sigs.k8s.io/autoscaler/cluster-autoscaler/apis`

where one of the folders is `provisioningrequest`.

Comment (alculquicondor) [reply cluster-autoscaler/provisioningrequest/apis/go.mod:1]: The reason being that we might have an generic (independent of the cloud provider) implementation of `provisioningrequest` and it might have dependencies that are not worth leaking to users of the API.

**Comment by tenzen-y** [comment]
> I don't think it's worth maintaining 3 packages. Maybe the client and APIs can be in the same package?

It makes sense.

File: cluster-autoscaler/provisioningrequest/apis/go.mod
Comment (tenzen-y) [reply cluster-autoscaler/provisioningrequest/apis/go.mod:1]: I see. Maybe, we can put a client folder on `sigs.k8s.io/autoscaler/cluster-autoscaler/apis/provisioningrequest/client` as well.

Comment (x13n) [reply cluster-autoscaler/provisioningrequest/apis/go.mod:1]: Doesn't `sigs.k8s.io` resolve to kubernetes-sigs org? This repo is in main kubernetes org, so shouldn't it be under `k8s.io/autoscaler/cluster-autoscaler`?

Comment (tenzen-y) [reply cluster-autoscaler/provisioningrequest/apis/go.mod:1]: Oops. Sorry, you're right.

File: cluster-autoscaler/boilerplate/boilerplate.generatego.txt
Comment (tenzen-y) [inline cluster-autoscaler/boilerplate/boilerplate.generatego.txt:1]: I added this file since there are no templates.

**Comment by tenzen-y** [comment]
Through some investigation, I found that `k8s.io/autoscaler/cluster-autoscaler/apis` pkg can not have `vendor` separate from `k8s.io/autoscaler/cluster-autoscaler`'s one due to the following error:

@x13n Do we need to have `vendor` for the `k8s.io/autoscaler/cluster-autoscaler/apis`?

> # k8s.io/autoscaler/cluster-autoscaler/provisioningrequest/provreqwrapper
> Error: provisioningrequest/provreqwrapper/wrapper.go:64:9: cannot use pr.v1Beta1PR.CreationTimestamp (variable of type "k8s.io/autoscaler/cluster-autoscaler/apis/vendor/k8s.io/apimachinery/pkg/apis/meta/v1".Time) as "k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/apimachinery/pkg/apis/meta/v1".Time value in return statement
> Error: provisioningrequest/provreqwrapper/wrapper.go:69:9: cannot use pr.v1Beta1PR (variable of type *v1beta1.ProvisioningRequest) as "k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/apimachinery/pkg/runtime".Object value in return statement: *v1beta1.ProvisioningRequest does not implement "k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/apimachinery/pkg/runtime".Object (wrong type for method DeepCopyObject)
> 		have DeepCopyObject() "k8s.io/autoscaler/cluster-autoscaler/apis/vendor/k8s.io/apimachinery/pkg/runtime".Object
> 		want DeepCopyObject() "k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/apimachinery/pkg/runtime".Object
> Error: provisioningrequest/provreqwrapper/wrapper.go:85:9: cannot use pr.v1Beta1PR.UID (variable of type "k8s.io/autoscaler/cluster-autoscaler/apis/vendor/k8s.io/apimachinery/pkg/types".UID) as "k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/apimachinery/pkg/types".UID value in return statement
> Error: provisioningrequest/provreqwrapper/wrapper.go:90:9: cannot use pr.v1Beta1PR.Status.Conditions (variable of type []"k8s.io/autoscaler/cluster-autoscaler/apis/vendor/k8s.io/apimachinery/pkg/apis/meta/v1".Condition) as []"k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/apimachinery/pkg/apis/meta/v1".Condition value in return statement
> Error: provisioningrequest/provreqwrapper/wrapper.go:95:35: cannot use conditions (variable of type []"k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/apimachinery/pkg/apis/meta/v1".Condition) as []"k8s.io/autoscaler/cluster-autoscaler/apis/vendor/k8s.io/apimachinery/pkg/apis/meta/v1".Condition value in assignment

https://github.com/kubernetes/autoscaler/actions/runs/7267376813/job/19801050209?pr=6315

Once removing vendor for the `k8s.io/autoscaler/cluster-autoscaler/apis` pkg, CI should work fine: https://github.com/kubernetes/autoscaler/actions/runs/7267869755/job/19802584596?pr=6315

**Comment by tenzen-y** [comment]
@x13n friendly ping :)

**Comment by tenzen-y** [comment]
I resolved conflicts.

**Comment by x13n** [comment]
Apologies for slow response.

/lgtm
/approve

@kisieland FYI

**Comment by tenzen-y** [comment]
/assign @mwielgus 
for `hack/for-go-proj.sh`

File: cluster-autoscaler/apis/go.mod
Comment (alculquicondor) [inline cluster-autoscaler/apis/go.mod:64]: Do we need these replace?

Comment (tenzen-y) [reply cluster-autoscaler/apis/go.mod:64]: I followed the go module in the following:

https://github.com/kubernetes/autoscaler/blob/d29ffd03b961b0cdd7b2f547035db9548c4ba022/cluster-autoscaler/go.mod#L211-L216

If this replacement is not needed, we can remove those.
@x13n WDYT?

Comment (alculquicondor) [reply cluster-autoscaler/apis/go.mod:64]: I think these `replace` could cause anyone using this package to have issues.

Also, maybe we can use released v0.29 versions instead.

Comment (tenzen-y) [reply cluster-autoscaler/apis/go.mod:64]: > I think these replace could cause anyone using this package to also have to replace.

That makes sense.

> Also, maybe we can use released v0.29 versions instead.

Indeed, the apis package could use the released version since that package doesn't depend on any k/k codes.

**Comment by tenzen-y** [comment]
I squashed old commits into one, and then I replaced alpha k8s libraries with the released ones.

@x13n @alculquicondor Please take another look.

**Comment by alculquicondor** [comment]
go.mod LGTM, in the sense that it will be usable from Kueue. I'll left the rest to the owners :)

**Comment by x13n** [comment]
Hm... This makes me think about a couple of things:
- Versioning should be consistent between main CA module and the new API module. Specifically, the k8s dependency should use the same tag.
- There's no binary that will be released for the new module, but we should keep the deps in sync going forwards. Do we need any changes to https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/hack/update-vendor.sh for this?
- Since 1.29 was already released, I think we should use `v1.30.0-alpha.0` on the master branch.

**Comment by tenzen-y** [comment]
> There's no binary that will be released for the new module, but we should keep the deps in sync going forwards. Do we need any changes to https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/hack/update-vendor.sh for this?

Yes, we need an additional script or modified `hack/update-vendor.sh`. I think another script would be better since the `cluster-autoscaler/hack/apis` doesn't have vendor.

**Comment by x13n** [comment]
> > Versioning should be consistent between main CA module and the new API module. Specifically, the k8s dependency should use the same tag.
> > Since 1.29 was already released, I think we should use v1.30.0-alpha.0 on the master branch.
> 
> These two points are in conflict. Users of the CA APIs shouldn't have to depend on alpha k8s modules.
> 
> I think using 1.29.x in the APIs and 1.30.0-alpha.y would work in most scenarios. CA code would compile with 1.30, as that satisfies both dependencies.

That makes sense, but will be inconsistent with tags unless we start releasing the API module independently from CA? Unless we follow something like this:
- master branch - API module uses latest k/k release deps, CA uses latest k/k alpha
- release branch 1.XX - both API module and CA use latest k/k patch: 1.XX.YY

Does that make sense to you? This should be documented in our release docs. CC @gjtempleton @MaciekPytel for thoughts.

> Yes, we need an additional script or modified `hack/update-vendor.sh`. I think another script would be better since the `cluster-autoscaler/hack/apis` doesn't have vendor.

Honestly I'm not sure if we really need CA to have vendor either, we could rename the script to `update-deps.sh` so it doesn't imply vendoring.

**Comment by tenzen-y** [comment]
> That makes sense, but will be inconsistent with tags unless we start releasing the API module independently from CA? Unless we follow something like this:
> 
> master branch - API module uses latest k/k release deps, CA uses latest k/k alpha
> release branch 1.XX - both API module and CA use latest k/k patch: 1.XX.YY
> Does that make sense to you? This should be documented in our release docs. CC @gjtempleton @MaciekPytel for thoughts.

@x13n I imagined that this (autoscaler) repository would cut releases separately from `cluster-autoscaler-1.x.y` like `cluster-autoscaler/apis`. If this repo wouldn't cut dedicated releases, we can not import `cluster-autoscaler/apis` to external projects using the tagging version due to Golang module restriction.

cc: @alculquicondor

**Comment by tenzen-y** [comment]
> > Yes, we need an additional script or modified hack/update-vendor.sh. I think another script would be better since the cluster-autoscaler/hack/apis doesn't have vendor.
> 
> Honestly I'm not sure if we really need CA to have vendor either, we could rename the script to update-deps.sh so it doesn't imply vendoring.

I'm ok with either way. I'll try to modify the existing script.

**Comment by x13n** [comment]
If we need separate process for cutting releases of the new module, we need to document it as well. I was thinking it could be piggybacking on the existing CA release tags though. Why is this an issue? We have 1:1 mappings between CA minor, k8s minor and go version used - does that render the api module unusable? I'd think if it depends on a specific k/k version, it is ok to rely on Go version used by the k/k as well.

**Comment by elmiko** [comment]
i haven't studied this PR deeply yet, but i agree with the thinking that it would be nice if these API pieces could be included from other projects that might be implementing common autoscaling components. thanks for the work here!

**Comment by x13n** [comment]
Yes, that makes sense to me, thanks. While reading about this a bit more, I came across https://go.dev/wiki/Modules#faqs--multi-module-repositories (see `Is it possible to add a module to a multi-module repository?` section). I think this is exactly the case we have here? Won't this PR lead to “ambiguous import” error when trying to use the new apis module?

**Comment by x13n** [comment]
The more I read about this, the more I wonder if it wouldn't be a better idea to put this module in a separate repo under kubernetes-sigs.

**Comment by alculquicondor** [comment]
That has the disadvantage that PRs that need API updates will have to be split among 2 repositories. The best model is what we have for k/k with the staging folder, but that sounds like overkill for just one submodule.

We might be better off doing a submodule in the repo. It's just 1 or 2 extra steps when branching a new version.

**Comment by x13n** [comment]
Yeah, I agree trying to follow in k/k footsteps here would be an overkill. Submodule is ok, just need to follow these extra steps. I see right now CA repo depends on `k8s.io/autoscaler/cluster-autoscaler/apis v0.0.0-00010101000000-000000000000`. What are the next steps after this PR is merged?

**Comment by tenzen-y** [comment]
> Yeah, I agree trying to follow in k/k footsteps here would be an overkill. Submodule is ok, just need to follow these extra steps. I see right now CA repo depends on `k8s.io/autoscaler/cluster-autoscaler/apis v0.0.0-00010101000000-000000000000`. What are the next steps after this PR is merged?

@x13n We need to work on https://github.com/kubernetes/autoscaler/pull/6315#issuecomment-1884726098 in this PR.
Please give me some time.

I think that we don't have any steps after this PR is merged. Any concerns?

**Comment by tenzen-y** [comment]
I have rebased. I'm still working on scripts.

**Comment by tenzen-y** [comment]
@x13n I modified the `update-deps.sh`. PTAL, thanks.

**Comment by tenzen-y** [comment]
@x13n friendly ping :)

File: cluster-autoscaler/hack/update-deps.sh
Comment (x13n) [inline cluster-autoscaler/hack/update-deps.sh:55]: I think `MODS` depend on `pkg`, shouldn't `MOD` definition be moved inside the inner loop here? Also, perhaps move to a function to avoid too much nesting.

**Review by x13n — COMMENTED** [review]
Thanks, just one minor comment, otherwise should be good to go.

Approving the PR, anyone can then lgtm & unhold later in case I won't (I'm about to go on vacation for a week).

/approve
/hold

File: cluster-autoscaler/hack/update-deps.sh
Comment (tenzen-y) [reply cluster-autoscaler/hack/update-deps.sh:55]: > I think MODS depend on pkg

@x13n I think it's not correct since `MODS` is set based on the `staging` directory in the k/k repository.
So, we can not move the `MOD` definition inside the inner loop.

**Comment by tenzen-y** [comment]
I have rebased this PR to resolve conflicts.

**Comment by tenzen-y** [comment]
Hi @mwielgus! Could you take a look this PR since we need to get approve from you due to changing of hack/for-go-proj.sh.
Thanks.

File: cluster-autoscaler/hack/update-deps.sh
Comment (x13n) [reply cluster-autoscaler/hack/update-deps.sh:55]: `MODS` is a result of catting `go.mod`, which is different for each `PKG`. For instance, `./go.mod` uses `k8s.io/component-base`, but `./apis/go.mod` doesn't.

Comment (tenzen-y) [reply cluster-autoscaler/hack/update-deps.sh:55]: Oh, it makes sense.
I'll try to fix it!

Comment (tenzen-y) [reply cluster-autoscaler/hack/update-deps.sh:55]: @x13n Indeed, the `MODS` is the result of catting `go.mod`, but we cat the `go.mod` after we perform "pushed ${REPO}". It means that we cat the go.mod in k/k.

So, I think that the current script is correct. WDYT?

```
pushd ${REPO}
git fetch --depth 1 origin v${VERSION}
git checkout FETCH_HEAD
MODS=($(
    cat go.mod | sed -n 's|.*k8s.io/\(.*\) => ./staging/src/k8s.io/.*|k8s.io/\1|p'
))
```

File: cluster-autoscaler/hack/update-deps.sh
Comment (tenzen-y) [inline cluster-autoscaler/hack/update-deps.sh:1]: @x13n As you proposed in another thread, I cleaned up this script. 
We can verify this script with `./hack/update-deps.sh ${K/K_ALPHA_VERSION} ${STABLE_APIS_VERSION}` like `./hack/update-deps.sh v1.30.0-alpha.3 v1.29.2`.

File: cluster-autoscaler/hack/update-deps.sh
Comment (tenzen-y) [reply cluster-autoscaler/hack/update-deps.sh]: Anyway, I cleaned up this script: https://github.com/kubernetes/autoscaler/pull/6315#discussion_r1518672607

**Comment by x13n** [comment]
Thanks! LGTM from my perspective.

/lgtm
/approve

**Comment by tenzen-y** [comment]
@mwielgus @maciekpytel @gjtempleton Could you get a chance to review this PR?

**Review by mwielgus — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
